README.md
# Pusher Channels Objective-C Client

[![Build Status](https://travis-ci.org/pusher/libPusher.png)](https://travis-ci.org/pusher/libPusher)

libPusher is an Objective-C client library for the [Pusher Channels](https://pusher.com/channels) realtime service.

Pusher Channels is a hosted service that sits between your web application and the browser that lets you deliver events in realtime using WebSockets.

The libPusher API mirrors the Pusher Channels JavaScript client as closely as possible, with some allowances for Objective-C conventions. In particular, whilst the JavaScript client uses event binding for all event handling, where events are pre-defined, libPusher uses the standard Cocoa delegation pattern.

[API Documentation](http://cocoadocs.org/docsets/libPusher/1.6/)

## Example
Subscribe to the ```chat``` channel and bind to the ```new-message``` event.

```objc
// self.client is a strong instance variable of class PTPusher
self.client = [PTPusher pusherWithKey:@"YOUR_API_KEY" delegate:self encrypted:YES];

// subscribe to channel and bind to event
PTPusherChannel *channel = [self.client subscribeToChannelNamed:@"chat"];
[channel bindToEventNamed:@"new-message" handleWithBlock:^(PTPusherEvent *channelEvent) {
	// channelEvent.data is a NSDictianary of the JSON object received
    NSString *message = [channelEvent.data objectForKey:@"text"];
	NSLog(@"message received: %@", message);
}];

[self.client connect];
```

## Supported platforms
### Deployment targets
- iOS 6.0 and above
- macOS (OS X) 10.9 and above

## Installation

Install using CocoaPods is recommended.

```ruby
pod 'libPusher', '~> 1.6.4'
```

Import Pusher into the class that wants to make use of the library.

```objc
#import <Pusher/Pusher.h>
```

If you want to use the ReactiveExtensions version of libPusher, add the following line to your Podfile instead.

```ruby
pod 'libPusher/ReactiveExtensions', '~> 1.6.4'
```

This will also load the core libPusher library and ReactiveCocoa as a dependency.

If you are not using CocoaPods, you can simply drop the extensions into your project.

## Usage

**Note**: in the following examples, ```client``` is a strong property. The instance returned by the ```pusherWithKey:*:``` methods will be auto-released, according to standard Objective-C return conventions. You must retain the client otherwise it will be auto-released before anything useful happens causing silent failures and unexpected behaviour.

### Create a client and connecting

```objc
self.client = [PTPusher pusherWithKey:@"YOUR_API_KEY" delegate:self encrypted:YES];

[self.client connect];
```

Note that clients do not connect automatically (as of version 1.5). You are responsible for calling connect as needed.

It is recommended to implement the PTPusherDelegate protocol in order to be notified when significant connection events happen such as connection errors, disconnects, and retries.

When the Pusher app is created in a different cluster to the default cluster, you must specify the `cluster` parameter.

```objc
self.client = [PTPusher pusehrWithKey:@"YOUR_API_KEY" delegate:self encrypt:YES cluster:@"YOUR_APP_CLUSTER"]

[self.client connect];
```

----------

### Subscribe to channels

Channels are a way of filtering the events you want your application to receive. In addition, private and presence channels can be used to control access to events and in the case of presence channels, see who else is subscribing to events. For more information on channels, see the [documentation](https://pusher.com/docs/channels/using_channels/channels).

There is no need to wait for the client to establish a connection before subscribing. You can subscribe to a channel immediately and any subscriptions will be created once the connection has connected.

#### Subscribing to public channels

```objc
PTPusherChannel *channel = [self.client subscribeToChannelNamed:@"chat"];
```

#### Subscribing to private channels

This method will add the appropriate ```private-``` prefix to the channel name for you and return a channel cast to the correct PTPusherChannel subclass PTPusherPrivateChannel.

Subscribing to private channels needs server-side authorization. See section [Channel authorization](#channel-authorization) for details.

```objc
// subscribe to private-chat channel
PTPusherPrivateChannel *private = [self.client subscribeToPrivateChannelNamed:@"chat"];
```

#### Subscribing to presence channels

This method will add the appropriate ```presence-``` prefix to the channel name for you and return a channel cast to the correct PTPusherChannel subclass PTPusherPresenceChannel.

Subscribing to presence channels needs server-side authorization. See section [Channel Authorization](#channel-authorization) for details.

```objc
// subscribe to presence-chat channel
PTPusherPresenceChannel *presence = [client subscribeToPresenceChannelNamed:@"chat" delegate:self];
```

It is recommended to implement ```PTPusherPresenceChannelDelegate``` protocol, to receive notifications for members subscribing or unsubscribing from the presence channel.

### Accessing subscribed channels

You can use the `channelNamed:` method to retrieve an existing subscribed channel. If you have not subscribed to the requested channel, it will return `nil`.

```objc
// get the 'chat' channel that you've already subscribed to
PTPusherChannel *channel = [self.client channelNamed:@"chat"];
```

### Unsubscribe from channels

If you no longer want to receive events on a given channel, you can unsubscribe from it.

```objc
PTPusherChannel *channel = [self.client channelNamed:@"chat"];
[channel unsubscribe];
```

### Channel object lifetime

When the client disconnects, all subscribed channels are implicitly unsubscribed (`isSubscribed` will return `NO`), however the channel objects will persist and so will any event bindings.

When the client reconnects, all previously subscribed channels will be resubscribed (which might involve another authentication request for any private/presence channels) and your existing event bindings will continue working as they did prior to the disconnection.

If you explicitly unsubscribe from a channel, **all event bindings will be removed and the client will remove the channel object from its list of subscribed channels**. If no other code has a strong reference to the channel object, it will be deallocated. If you resubscribe to the channel, a new channel object will be created. You should bear this in mind if you maintain any strong references to a channel object in your application code.

### Channel authorization

Private and presence channels require server-side authorization before they can connect.

**Note**: Make sure your server responds correctly to the authentication request. See the [authentication signature](https://pusher.com/docs/channels/library_auth_reference/auth-signatures) and [user authentication](https://pusher.com/docs/channels/server_api/authenticating-users) docs for details and examples on how to implement authorization on the server side.

In order to connect to a private or presence channel, you first need to configure your server authorization URL.

```objc
self.client.authorizationURL = [NSURL URLWithString:@"https://www.yourserver.com/authorize"];
```

When you attempt to connect to a private or presence channel, libPusher will make a form-encoded POST request to the above URL, passing along the ```socket_id``` and ```channel_name``` as parameters. Prior to sending the request, the Pusher delegate will be notified, passing in the channel and the NSMutableURLRequest instance that will be sent.

#### Custom authorization

In order to perform custom authorization, you need to assign the `channelAuthorizationDelegate` and implement the `PTPusherChannelAuthorizationDelegate` protocol, which currently contains a single method. The following example uses the `AFNetworking` library to perform server-based authorization:

```objc
- (BOOL)applicationDidFinishLaunching:(UIApplication *)application
{
  ...
  self.pusherClient.channelAuthorizationDelegate = self;
}

- (void)pusherChannel:(PTPusherChannel *)channel requiresAuthorizationForSocketID:(NSString *)socketID completionHandler:(void(^)(BOOL isAuthorized, NSDictionary *authData, NSError *error))completionHandler
{
  NSDictionary *authParameters = :@{@"socket_id": socketID, @"channel_name": channel.name};

  [[MyHTTPClient sharedClient] postPath:@"/pusher/auth" parameters:authParameters success:^(AFHTTPRequestOperation *operation, id responseObject) {
      completionHandler(YES, responseObject, nil);
  } failure:^(AFHTTPRequestOperation *operation, NSError *error) {
    completionHandler(NO, nil, error);
  }];
}
```

----------

### Binding to events

There are generally two ways to bind to events: binding to an event on the PTPusher client itself or binding to a specific channel.

Two types of direct binding are supported: target/action and block-based bindings. The examples below using block-based bindings.

#### Binding to a channel

Once you have created an instance of PTPusherChannel, you can set up event bindings. There is no need to wait for the PTPusher client connection to be established or the channel to be subscribed.

When you bind to events on a single channel, you will only receive events with that name if they are sent over this channel.

```objc
PTPusherChannel *channel = [self.client subscribeToChannelNamed:@"chat"];
[channel bindToEventNamed:@"new-message" handleWithBlock:^(PTPusherEvent *channelEvent) {
  // channelEvent.data is a NSDictionary of the JSON object received
}];
```

#### Binding directly to the client

When you bind to events on the client, you will receive all events with that name, regardless of the channel from which they originated.

```objc
[self.client bindToEventNamed:@"new-message" handleWithBlock:^(PTPusherEvent *event) {
  // event.data is a NSDictionary of the JSON object received
}];
```

#### Remove bindings

If you no longer want to receive events with a specific name, you can remove the binding. Removing a binding is as simple as storing a reference to the binding object, then passing that as an argument to ```removeBinding:``` at a later point.

**Note:** Binding objects are owned by the client or channel that they relate to and will exist for the lifetime of the binding. For this reason, you generally only need to store a weak reference to the binding object in order to remove the binding. In the event that something else removes the binding (perhaps as a result of calling `removeAllBindings` or explicitly unsubscribing from the channel), the weak reference will ensure that the binding object will become nil, so you should check this before calling `removeBinding:`.

```objc
// _binding is a weak reference
_binding = [self.client bindToEventNamed:@"new-message" target:self action:@selector(handleEvent:)];

// check that nothing else has removed the binding already
if (_binding) {
  [self.client removeBinding:_binding];
}
```

#### Memory management considerations for block-based bindings

Similar caveats apply to block-based bindings as they do to using block based `NSNotificationCenter` observers, i.e. when referencing `self` in your event handler block, it is possible in some situations to create retain cycles or prevent `self` from being deallocated.

When you reference `self` in your event handler block, the block will retain a strong reference to `self`. This means that `self` will never be deallocated until the binding (and in turn the event handler block) is destroyed by removing the binding. For this reason, you should be wary about removing event bindings in `dealloc` methods as `dealloc` will never be called if the binding references `self`.

For example, you might push a `UIViewController` on to a `UINavigationController` stack, then create an event binding in that view controller's `viewDidAppear:` method:

```objc
- (void)viewDidAppear:(BOOL)animated
{
  // _binding is a weak instance variable
  _binding = [self.client bindToEventNamed:@"new-message" handleWithBlock:^(PTPusherEvent *event) {
    [self doSomethingWithEvent:event];
  }];
}
```

If you were to then pop that view controller off the navigation stack without removing the event binding, because the binding block has a strong reference to `self`, the view controller will never be deallocated and you will have a memory leak.

You can handle this in one of two ways. The first is to ensure you remove the binding when in the corresponding `viewDidDisappear:`

```objc
- (void)viewDidDisappear:(BOOL)animated
{
  [self.client removeBinding:_binding];
}
```

The second, is to prevent a strong reference to `self` being captured in the first place:

```objc
- (void)viewDidAppear:(BOOL)animated
{
  __weak typeof(self) weakSelf = self;

  // _binding is a weak instance variable
  _binding = [self.client bindToEventNamed:@"new-message" handleWithBlock:^(PTPusherEvent *event) {
    __strong typeof(weakSelf) strongSelf = weakSelf;
    [strongSelf doSomethingWithEvent:event];
  }];
}
```

Finally, if you reference `self` in a block and store a *strong* reference to the binding object, you will create a retain cycle (`self` -> `binding` -> `block` -> `self`). You should avoid keeping strong references to binding objects but if you really need to, you should ensure you only capture a weak reference to `self` in the block as in the above example.

#### Binding to all events

In some cases it might be useful to bind to all events of a client or channel.

CHANGES.md
# 1.6.4 - January 2020

* Fix SocketRocket header imports (#272)

# 1.6.3 - March 2017

* Fix `PTURLRequestOperation` memory leak (#246)
* Ensure socket IDs still match before proceeding with subscription (#237)

# 1.6.2 - September 2016

* Added option to unsubscribe from all channels without knowing their names (#200)
* PTPusherEventReceivedNotification gets posted at correct point (#192)
* NSURLSession replaced NSURLConnection
* OHHTTPStubs to version 5.1.0
* Check and cast presence channel user id to NSString (#193)
* Expose channels reference to see what channels are currently subscribed (#197)

# 1.6.1 - September 2015

* Update SocketRocket to 0.4.1 (#181)
* Update Gemfile to use more recent version of Cocoapods
* Changed iOS deployment target to 6.0

# 1.6 - March 2015

* Swift and 64-bit compatibility
* Added new delegate method to support extracting auth response from wrapped payloads (#171)
* Removed deprecated methods and delegate calls
* Removed SocketRocket from the public interface (#157)
* Fix deprecation warning for willAuthorizeChannelWithRequest (#147)
* Use NSDate instead of C API for setting timestamp on auth request (#133)

# 1.5 - December 2013

This release contains some significant bug fixes and API changes. All deprecated API in this release will be removed in the next release after this one.

## Upgrade notes

This is a summary of the changes in this release and notes on how to upgrade. The recommended way of installing libPusher is to use CocoaPods.

As of this version, support for iOS < 5.0 and OSX < 10.8 has been dropped.

### Connection is no longer automatic

`PTPusher` will no longer connect automatically on initialisation and all methods that accept a connect `connectAutomatically` parameter (including the initialiser and all factory methods) are deprecated. 

You should now explicitly call `connect` when you are ready to connect.

The method `reconnectAutomatically` has been removed completely (see below).

### Improvements to disconnection handling

`PTPusher` now goes to great lengths to ensure it remains connected whenever possible, including correctly handling error codes returned by the Pusher service (see http://pusher.com/docs/pusher_protocol#error-codes). 

In most cases where the connection fails or disconnects, Pusher will attempt to reconnect either immediately or after a configured delay (which defaults to 5s and can still be customised using the `reconnectDelay` property. 

In the case where Pusher returns a 4100 error code (over capacity), reconnection attempts will be attempted using a linear back-off delay. In all cases, Pusher will limit reconnection attempts to a maximum of three attempts before giving up.

There are two circumstances in which the client will not reconnect automatically:

  * The client disconnects with an error code in the 4000-4099 range, which generally indicates a client misconfiguration.
  * The client is not able to connect in the first place (i.e. it fails on it's first attempt), which normally indicates that there is no network connection or there is a server-side issue.

A new delegate method, `pusher:connection:didDisconnectWithError:willAttemptReconnect:` will be called when Pusher disconnects. The `willAttemptReconnect:` parameter will indicate to clients that `PTPusher` will automatically reconnect. If this is `NO`, clients should decide how they want to proceed (e.g. you may want to check network reachability and manually reconnect when reachability changes).

These changes should ensure that libPusher is much more reliable. You no longer need to explicitly disconnect and reconnect the client when going to the background or when the device is locked - in these situations the client will automatically reconnect when the app returns to the foreground or the device is unlocked.

The delegate methods `pusher:connection:didDisconnectWithError` and `pusher:connection:didDisconnect:` are now deprecated and will be removed in the next release.

Additionally, the `reconnectAutomatically` property of `PTPusher` has been deprecated. Setting this will not affect the automatic reconnection behaviour of `PTPusher`. If you need to take control of the auto-reconnect behaviour, a new delegate method is provided, `pusher:connectionWillAutomaticallyReconnect:afterDelay:`. Returning `NO` from this method will prevent the automatic reconnection attempt from happening. You will be responsible for manually reconnecting the client.

Pusher can be prevented from connecting *at any time* by returning `NO` from another new delegate method, `pusher:connectionWillConnect:`.

### Changes to presence channel members API

The API for accessing members of a channel has been brought in line with the JavaScript client.

Presence channels have a property `members`, which returns a instance of `PTPusherChannelMembers`, which is an unordered collection of members. Members can be retrieved by ID using the `memberWithID:` method. Members are represented by instances of the class `PTPusherChannelMember` rather than `NSDictionary` - see the headers for more information. There is also a property, `me`, which returns your own member object.

### Experimental ReactiveExtensions

This release contains some extensions that allow binding to events using ReactiveCocoa signals. These extensions are bundled as a separate library and if you're using CocoaPods, a sub-spec that is excluded from the default spec. These are still experimental so proceed with caution. You can add them to your project by adding `libPusher/ReactiveExtensions` to your `Podfile`. 

### Bug Fixes

* There have been numerous bug fixes around connection handling and Pusher should generally be more stable and remain connected in most cases where you have network connectivity. 
* libPusher now uses socket-native ping/pong and also sends client-side pings to keep the connection alive.

Library/PTPusher.m
//
//  PTPusher.m
//  PusherEvents
//
//  Created by Luke Redpath on 22/03/2010.
//  Copyright 2010 LJR Software Limited. All rights reserved.
//

#import "PTPusher.h"
#import "PTEventListener.h"
#import "PTPusherEvent.h"
#import "PTPusherChannel.h"
#import "PTPusherEventDispatcher.h"
#import "PTTargetActionEventListener.h"
#import "PTBlockEventListener.h"
#import "PTPusherErrors.h"
#import "PTPusherChannelServerBasedAuthorization.h"
#import "PTPusherChannel_Private.h"
#import <SocketRocket/SRWebSocket.h>

#define kPUSHER_HOST @"ws.pusherapp.com"

typedef NS_ENUM(NSUInteger, PTPusherAutoReconnectMode) {
  PTPusherAutoReconnectModeNoReconnect,
  PTPusherAutoReconnectModeReconnectImmediately,
  PTPusherAutoReconnectModeReconnectWithConfiguredDelay,
  PTPusherAutoReconnectModeReconnectWithBackoffDelay
};

NSURL *PTPusherConnectionURL(NSString *host, NSString *key, NSString *clientID, BOOL secure);

NSString *const PTPusherEventReceivedNotification = @"PTPusherEventReceivedNotification";
NSString *const PTPusherEventUserInfoKey          = @"PTPusherEventUserInfoKey";
NSString *const PTPusherErrorDomain               = @"PTPusherErrorDomain";
NSString *const PTPusherFatalErrorDomain          = @"PTPusherFatalErrorDomain";
NSString *const PTPusherErrorUnderlyingEventKey   = @"PTPusherErrorUnderlyingEventKey";

/** The Pusher protocol version, used to determined which features
 are supported.
 */
#define kPTPusherClientProtocolVersion 6

NSURL *PTPusherConnectionURL(NSString *host, NSString *key, NSString *clientID, BOOL encrypted)
{
  NSString *scheme = ((encrypted == YES) ? @"wss" : @"ws");
  NSString *URLString = [NSString stringWithFormat:@"%@://%@/app/%@?client=%@&protocol=%d&version=%@",
                         scheme, host, key, clientID, kPTPusherClientProtocolVersion, kPTPusherClientLibraryVersion];
  return [NSURL URLWithString:URLString];
}

#define kPTPusherDefaultReconnectDelay 5.0

@interface PTPusher ()
@property (nonatomic, strong, readwrite) PTPusherConnection *connection;
@property (nonatomic, assign) PTPusherAutoReconnectMode autoReconnectMode;
@end

#pragma mark -

@implementation PTPusher {
  NSUInteger _numberOfReconnectAttempts;
  NSUInteger _maximumNumberOfReconnectAttempts;
  PTPusherEventDispatcher *dispatcher;
  NSMutableDictionary *channels;
  PTPusherChannelServerBasedAuthorization *serverAuthorizationStrategy;
}

@synthesize connection = _connection;
@synthesize delegate;

- (id)initWithConnection:(PTPusherConnection *)connection
{
  if (self = [super init]) {
    dispatcher = [[PTPusherEventDispatcher alloc] init];
    channels = [[NSMutableDictionary alloc] init];

    self.connection = connection;
    self.connection.delegate = self;
    self.reconnectDelay = kPTPusherDefaultReconnectDelay;

    /* Three reconnection attempts should be more than enough attempts
     * to reconnect where the user has simply locked their device or
     * backgrounded the app.
     *
     * If there is no internet connection, we will only end up retrying
     * once as after the first failure we will no longer auto-retry.
     *
     * We may consider making this user-customisable in future but not
     * for now.
     */
    _maximumNumberOfReconnectAttempts = 3;
  }
  return self;
}

+ (instancetype)pusherWithKey:(NSString *)key delegate:(id<PTPusherDelegate>)delegate
{
  return [self pusherWithKey:key delegate:delegate encrypted:YES];
}

+ (instancetype)pusherWithKey:(NSString *)key delegate:(id<PTPusherDelegate>)delegate encrypted:(BOOL)isEncrypted
{
  return [self pusherWithKey:(NSString *)key delegate:(id<PTPusherDelegate>)delegate encrypted:(BOOL)isEncrypted cluster:(NSString *) nil];
}

+ (instancetype)pusherWithKey:(NSString *)key delegate:(id<PTPusherDelegate>)delegate encrypted:(BOOL)isEncrypted cluster:(NSString *) cluster
{
    NSString * hostURL;
    if ([cluster length] == 0) {
        hostURL = kPUSHER_HOST;
    } else {
        hostURL = [NSString stringWithFormat:@"ws-%@.pusher.com", cluster];
    }

    NSURL *serviceURL = PTPusherConnectionURL(hostURL, key, @"libPusher", isEncrypted);
    PTPusherConnection *connection = [[PTPusherConnection alloc] initWithURL:serviceURL];
    PTPusher *pusher = [[self alloc] initWithConnection:connection];
    pusher.delegate = delegate;
    return pusher;
}

- (void)dealloc;
{
  [_connection setDelegate:nil];
  [_connection disconnect];
}

#pragma mark - Authorization strategy

- (void)setAuthorizationURL:(NSURL *)authorizationURL
{
  if (self.channelAuthorizationDelegate && ![self.channelAuthorizationDelegate isKindOfClass:[PTPusherChannelServerBasedAuthorization class]])
    return;

  serverAuthorizationStrategy = [[PTPusherChannelServerBasedAuthorization alloc] initWithAuthorizationURL:authorizationURL];

  __weak PTPusher *weakSelf = self;

  // use this to support our current delegate-based API for HTTP authorization
  [serverAuthorizationStrategy customizeOperationsWithBlock:^(PTPusherChannelAuthorizationOperation *op, PTPusherChannel *channel) {
    __strong PTPusher *strongSelf = weakSelf;

#pragma clang diagnostic push
#pragma clang diagnostic pop
    if ([strongSelf.delegate respondsToSelector:@selector(pusher:willAuthorizeChannel:withAuthOperation:)]) {
      [strongSelf.delegate pusher:strongSelf willAuthorizeChannel:channel withAuthOperation:op];
    }
  }];
  self.channelAuthorizationDelegate = serverAuthorizationStrategy;
}

- (NSURL *)authorizationURL
{
  return serverAuthorizationStrategy.authorizationURL;
}

#pragma mark - Connection management

- (void)setReconnectDelay:(NSTimeInterval)reconnectDelay
{
  _reconnectDelay = MAX(reconnectDelay, 1);
}

- (void)connect
{
  _numberOfReconnectAttempts = 0;
  self.autoReconnectMode = PTPusherAutoReconnectModeReconnectWithConfiguredDelay;
  [self.connection connect];
}

- (void)disconnect
{
  // we do not want to reconnect if a user explicitly disconnects
  self.autoReconnectMode = PTPusherAutoReconnectModeNoReconnect;
  [self.connection disconnect];
}

#pragma mark - Binding to events

- (PTPusherEventBinding *)bindToEventNamed:(NSString *)eventName target:(id)target action:(SEL)selector
{
  return [dispatcher addEventListenerForEventNamed:eventName target:target action:selector];
}

- (PTPusherEventBinding *)bindToEventNamed:(NSString *)eventName handleWithBlock:(PTPusherEventBlockHandler)block
{
  return [self bindToEventNamed:eventName handleWithBlock:block queue:dispatch_get_main_queue()];
}

- (PTPusherEventBinding *)bindToEventNamed:(NSString *)eventName handleWithBlock:(PTPusherEventBlockHandler)block queue:(dispatch_queue_t)queue
{
  return [dispatcher addEventListenerForEventNamed:eventName block:block queue:queue];
}

- (void)removeBinding:(PTPusherEventBinding *)binding
{
  [dispatcher removeBinding:binding];
}

- (void)removeAllBindings
{
  [dispatcher removeAllBindings];
}

#pragma mark - Subscribing to channels

- (PTPusherChannel *)subscribeToChannelNamed:(NSString *)name
{
  PTPusherChannel *channel = channels[name];
  if (channel == nil) {
    channel = [PTPusherChannel channelWithName:name pusher:self];
    channels[name] = channel;
  }
  // private/presence channels require a socketID to authenticate
  if (self.connection.isConnected && self.connection.socketID) {
    [self subscribeToChannel:channel];
  }
  return channel;
}

- (PTPusherPrivateChannel *)subscribeToPrivateChannelNamed:(NSString *)name
{
  return (PTPusherPrivateChannel *)[self subscribeToChannelNamed:[NSString stringWithFormat:@"private-%@", name]];
}

- (PTPusherPresenceChannel *)subscribeToPresenceChannelNamed:(NSString *)name
{
  return (PTPusherPresenceChannel *)[self subscribeToChannelNamed:[NSString stringWithFormat:@"presence-%@", name]];
}

- (PTPusherPresenceChannel *)subscribeToPresenceChannelNamed:(NSString *)name delegate:(id<PTPusherPresenceChannelDelegate>)presenceDelegate
{
  PTPusherPresenceChannel *channel = [self subscribeToPresenceChannelNamed:name];
  channel.presenceDelegate = presenceDelegate;
  return channel;
}

- (PTPusherChannel *)channelNamed:(NSString *)name
{
  return channels[name];
}

- (NSDictionary *)subscribedChannels
{
  return [[NSDictionary alloc] initWithDictionary:channels copyItems:NO];
}

/* This is only called when a client explicitly unsubscribes from a channel
 * by calling either [channel unsubscribe] or using the deprecated API
 * [client unsubscribeFromChannel:].
 *
 * This effectively ends the lifetime of a channel: the client will remove it
 * from it's channels collection and all bindings will be removed. If no other
 * code outside of libPusher has a strong reference to the channel, it will
 * be deallocated.
 *
 * This is different to implicit unsubscribes (where the connection has been lost)
 * where the channel will object will remain and be re-subscribed when connection
 * is re-established.
 *
 * A pusher:unsubscribe event will only be sent if there is a connection, otherwise
 * it's not necessary as the channel is already implicitly unsubscribed due to the
 * disconnection.
 */
- (void)__unsubscribeFromChannel:(PTPusherChannel *)channel
{
  NSParameterAssert(channel != nil);

  [channel removeAllBindings];

  if (self.connection.isConnected) {
    [self sendEventNamed:@"pusher:unsubscribe"
                    data:@{@"channel": channel.name}];
  }

  [channels removeObjectForKey:channel.name];

  if ([self.delegate respondsToSelector:@selector(pusher:didUnsubscribeFromChannel:)]) {
    [self.delegate pusher:self didUnsubscribeFromChannel:channel];
  }
}

- (void)subscribeToChannel:(PTPusherChannel *)channel
{
  if (channel.isPrivate) {
    NSString *socketID = self.connection.socketID;
    [self.channelAuthorizationDelegate pusherChannel:channel requiresAuthorizationForSocketID:socketID completionHandler:^(BOOL isAuthorized, NSDictionary *authData, NSError *error) {
      
      if (!self.connection.isConnected) return;
      if (self.connection.socketID != socketID) return; // Socket may have changed since the auth request was made
      
      if (isAuthorized) {
        [channel subscribeWithAuthorization:authData];
      }
      else {
        if (error == nil) {
          error = [NSError errorWithDomain:PTPusherErrorDomain code:PTPusherSubscriptionUnknownAuthorisationError userInfo:nil];
        }

        if ([self.delegate respondsToSelector:@selector(pusher:didFailToSubscribeToChannel:withError:)]) {
          [self.delegate pusher:self didFailToSubscribeToChannel:channel withError:error];
        }
      }
    }];
  }
  else {
    [channel subscribeWithAuthorization:nil];
  }
}

- (void)subscribeAll
{
  for (PTPusherChannel *channel in [channels allValues]) {
    [self subscribeToChannel:channel];
  }
}

- (void)unsubscribeAllChannels
{
  if (channels && channels.count) {
    for (PTPusherChannel *channel in [channels allValues]) {
      [channel unsubscribe];
    }
  }
}

#pragma mark - Sending events

- (void)sendEventNamed:(NSString *)name data:(id)data
{
  [self sendEventNamed:name data:data channel:nil];
}

- (void)sendEventNamed:(NSString *)name data:(id)data channel:(NSString *)channelName
{
  NSParameterAssert(name);

  if (self.connection.isConnected == NO) {
    NSLog(@"Warning: attempting to send event while disconnected. Event will not be sent.");
    return;
  }

  NSMutableDictionary *payload = [NSMutableDictionary dictionary];
  payload[PTPusherEventKey] = name;

  if (data) {
    payload[PTPusherDataKey] = data;
  }

  if (channelName) {
    payload[PTPusherChannelKey] = channelName;
  }
  [self.connection send:payload];
}

#pragma mark - PTPusherConnection delegate methods

- (BOOL)pusherConnectionWillConnect:(PTPusherConnection *)connection
{
  if ([self.delegate respondsToSelector:@selector(pusher:connectionWillConnect:)]) {
    return [self.delegate pusher:self connectionWillConnect:connection];
  }
  return YES;
}

- (void)pusherConnectionDidConnect:(PTPusherConnection *)connection
{
  _numberOfReconnectAttempts = 0;

  if ([self.delegate respondsToSelector:@selector(pusher:connectionDidConnect:)]) {
    [self.delegate pusher:self connectionDidConnect:connection];
  }

  [self subscribeAll];
}

- (void)pusherConnection:(PTPusherConnection *)connection didDisconnectWithCode:(NSInteger)errorCode reason:(NSString *)reason wasClean:(BOOL)wasClean
{
  NSError *error = nil;

  if (errorCode > 0 && errorCode != SRStatusCodeNormal && errorCode != SRStatusCodeGoingAway) {
    if (reason == nil) {
        reason = @"Unknown error"; // not sure what could cause this to be nil, but just playing it safe
    }

    NSString *errorDomain = PTPusherErrorDomain;

    if (errorCode >= 400 && errorCode <= 4099) {
      errorDomain = PTPusherFatalErrorDomain;
    }

    // check for error codes based on the Pusher Websocket protocol see http://pusher.com/docs/pusher_protocol
    error = [NSError errorWithDomain:errorDomain code:errorCode userInfo:@{@"reason": reason}];

    // 4000-4099 -> The connection SHOULD NOT be re-established unchanged.
    if (errorCode >= 4000 && errorCode <= 4099) {
      [self handleDisconnection:connection error:error reconnectMode:PTPusherAutoReconnectModeNoReconnect];
    } else
    // 4200-4299 -> The connection SHOULD be re-established immediately.
    if(errorCode >= 4200 && errorCode <= 4299) {
      [self handleDisconnection:connection error:error reconnectMode:PTPusherAutoReconnectModeReconnectImmediately];
    }

    else {
      // i.e. 4100-4199 -> The connection SHOULD be re-established after backing off.
      [self handleDisconnection:connection error:error reconnectMode:PTPusherAutoReconnectModeReconnectWithBackoffDelay];
    }
  }
  else {
    [self handleDisconnection:connection error:error reconnectMode:self.autoReconnectMode];
  }
}

- (void)pusherConnection:(PTPusherConnection *)connection didFailWithError:(NSError *)error wasConnected:(BOOL)wasConnected
{
  if (wasConnected) {
    [self handleDisconnection:connection error:error reconnectMode:PTPusherAutoReconnectModeReconnectImmediately];
  }
  else {
    if ([self.delegate respondsToSelector:@selector(pusher:connection:failedWithError:)]) {
      [self.delegate pusher:self connection:connection failedWithError:error];
    }
  }
}

- (void)pusherConnection:(PTPusherConnection *)connection didReceiveEvent:(PTPusherEvent *)event
{
  [[NSNotificationCenter defaultCenter]
     postNotificationName:PTPusherEventReceivedNotification
     object:self
     userInfo:@{PTPusherEventUserInfoKey: event}];

  if ([event isKindOfClass:[PTPusherErrorEvent class]]) {
    if ([self.delegate respondsToSelector:@selector(pusher:didReceiveErrorEvent:)]) {
      [self.delegate pusher:self didReceiveErrorEvent:(PTPusherErrorEvent *)event];
    }
  }

  if (event.channel) {
    [channels[event.channel] dispatchEvent:event];
  }
  [dispatcher dispatchEvent:event];
}

- (void)handleDisconnection:(PTPusherConnection *)connection error:(NSError *)error reconnectMode:(PTPusherAutoReconnectMode)reconnectMode
{
  for (PTPusherChannel *channel in [channels allValues]) {
    [channel handleDisconnect];
  }

  BOOL willReconnect = NO;

  if (reconnectMode > PTPusherAutoReconnectModeNoReconnect && _numberOfReconnectAttempts < _maximumNumberOfReconnectAttempts) {
    willReconnect = YES;
  }

  if ([self.delegate respondsToSelector:@selector(pusher:connection:didDisconnectWithError:willAttemptReconnect:)]) {
    [self.delegate pusher:self connection:connection didDisconnectWithError:error willAttemptReconnect:willReconnect];
  }

  if (willReconnect) {
    [self reconnectUsingMode:reconnectMode];
  }
}

#pragma mark - Private

- (void)reconnectUsingMode:(PTPusherAutoReconnectMode)reconnectMode
{
  _numberOfReconnectAttempts++;

  NSTimeInterval delay;

  switch (reconnectMode) {
    case PTPusherAutoReconnectModeReconnectImmediately:
      delay = 0;
      break;
    case PTPusherAutoReconnectModeReconnectWithConfiguredDelay:
      delay = self.reconnectDelay;
      break;
    case PTPusherAutoReconnectModeReconnectWithBackoffDelay:
      delay = self.reconnectDelay * _numberOfReconnectAttempts;
      break;
    default:
      delay = 0;
      break;
  }

  if ([self.delegate respondsToSelector:@selector(pusher:connectionWillAutomaticallyReconnect:afterDelay:)]) {
    BOOL shouldProceed = [self.delegate pusher:self connectionWillAutomaticallyReconnect:_connection afterDelay:delay];

    if (!shouldProceed) return;
  }

  dispatch_time_t popTime = dispatch_time(DISPATCH_TIME_NOW, delay * NSEC_PER_SEC);
  dispatch_after(popTime, dispatch_get_main_queue(), ^(void){
    [self.connection connect];
  });
}

@end
